Fixes #30466 - trigger toasts from API middleware

Ron-Lavi opened the pull request:
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->
## Toast handling

Instead of managing and dispatching the toast action by yourself, you can use the API middleware shortcuts,
by passing `toastSuccess` or `toastError` as keys in the API action, e.g:

```js
dispatch(
  get({
    key: MY_SPECIAL_KEY,
    url,
    toastError: error => `Oh no! Something went wrong while submiting the form, the server returned the following error: ${error}`,
    toastSuccess: response => `Yay! task: ${response.taskID} was successfully created.`
  })
);
```

![c7c586875109ec38f103e3fc9985c0b2](https://user-images.githubusercontent.com/26363699/88090852-da867800-cb96-11ea-9b6b-0df4947e25cb.png)

ronlavi2412 committed:
Fixes #30466 - trigger toasts from API middleware

Ron-Lavi commented on webpack/assets/javascripts/react_app/redux/API/APIRequest.js:
```diff
@@ -15,26 +16,37 @@ export const apiRequest = async (
 ) => {
   const { REQUEST, SUCCESS, FAILURE } = actionTypeGenerator(key, actionTypes);
   const modifiedPayload = { ...payload, url };
+
```
Added some spaces for readability

Ron-Lavi commented on webpack/assets/javascripts/react_app/redux/actions/toasts/index.js:
```diff
@@ -3,7 +3,7 @@ import uuidV1 from 'uuid/v1';
 
 export const addToast = toast => {
-  const key = uuidV1();
+  const key = toast.key || uuidV1();
```
it was one annoying `uuid` ( especially in tests )

amirfefer commented on webpack/assets/javascripts/react_app/redux/actions/common/forms.js:
```diff
@@ -85,6 +79,7 @@ export const submitForm = ({
         actionTypes,
         handleError,
         handleSuccess,
+        successToast,
```
does it make sense to add an error toast as well?

amirfefer commented on webpack/assets/javascripts/react_app/redux/API/APIActions.js:
```diff
@@ -10,8 +10,10 @@ const { GET, POST, PUT, DELETE, PATCH } = API_OPERATIONS;
- * @param { Object } payload.handleSuccess a success handling callback.
+ * @param { Function } payload.handleError an error handling callback.
+ * @param { Function } payload.handleSuccess a success handling callback.
+ * @param { String } payload.errorToast an error toast will be triggered with this message after API error.
+ * @param { String } payload.successToast a succes toast will be triggered with this message after API success.
```
can you add these new props to our documentation?

amirfefer reviewed (changes requested):
Thanks @laviro ! a great addition to our API stack  :+1: 
can you please extend `APIRequest`  test for including this new feature?

Ron-Lavi commented:
Thanks @amirfefer !
In the last commit:
- Added tests
- Added section to storybook
- Added also `errorToast` to `forms.js`
- Changed it into a function so the toast can contain info from the server, e.g:
```js
dispatch(
  get({
    key: MY_SPECIAL_KEY,
    url,
    toastError: error => `Oh no! Something went wrong while submiting the form, the server returned the following error: ${error}`,
    toastSuccess: response => `Yay! task: ${response.taskID} was successfully created.`
  })
);
```

amirfefer reviewed (approved):
Thanks @laviro ! 
works well

ready for merge after stabilization

amirfefer commented:
Merging, Thanks @laviro !
